### PR TITLE
Read approximation on SEC_E_INCOMPLETE_MESSAGE

### PR DESCRIPTION
**Pull request by steffengy**
Basic implementation to use the approximation returned by the crypto api,
to reduce the amount of API calls.

This essentially reads atleast the approximated amount of bytes.

### CODE REVIEW

File: src/tls_stream.rs [hunk 1]
```diff
@@ -541,17 +545,27 @@ impl<S> TlsStream<S>
...
+            let min_len = cmp::max(1024, self.needs_read + 2 * existing_len);
+            if self.enc_in.get_ref().len() < min_len {
+                self.enc_in.get_mut().resize(min_len, 0);
+            }
+            let nread = {
+                let buf = &mut self.enc_in.get_mut()[existing_len..];
+                try!(self.stream.read(buf))
+            };
+            self.enc_in.set_position((existing_len + nread) as u64);
+            self.needs_read = self.needs_read.checked_sub(nread).unwrap_or(0);
```
Comment (sfackler) [inline src/tls_stream.rs]: This can just be saturating_sub(nread)

File: src/tls_stream.rs [hunk 2]
```diff
@@ -595,19 +609,25 @@ impl<S> TlsStream<S>
...
+                        bufs[1].cbBuffer as usize
+                    } else {
+                        1
+                    };
                     Ok(())
                 }
-                state @ winapi::SEC_I_CONTEXT_EXPIRED |
-                state @ winapi::SEC_I_RENEGOTIATE => {
+                winapi::SEC_I_CONTEXT_EXPIRED => {
+                    self.shutdown()
```
Comment (sfackler) [inline src/tls_stream.rs]: This is the fix for that wikipedia issue you ran into, right?

File: src/tls_stream.rs [hunk 3]
```diff
@@ -541,17 +545,27 @@ impl<S> TlsStream<S>
...
     fn read_in(&mut self) -> io::Result<usize> {
-        let existing_len = self.enc_in.position() as usize;
-        let min_len = cmp::max(1024, 2 * existing_len);
-        if self.enc_in.get_ref().len() < min_len {
-            self.enc_in.get_mut().resize(min_len, 0);
+        let mut sum_nread = 0;
+
+        while self.needs_read > 0 {
+            let existing_len = self.enc_in.position() as usize;
+            let min_len = cmp::max(1024, self.needs_read + 2 * existing_len);
```
Comment (sfackler) [inline src/tls_stream.rs]: This seems a little suspicious - maybe it should be something like `max(max(1024, existing_len * 2), self.needs_read)`? We want to make sure our buffer is larger than 1023, larger than the amount we want to read, and is doubling in size past that point.

**Comment by sfackler** [comment]
Looks good to me other than a couple nits!

File: src/tls_stream.rs (hunk 2 above)
Comment (steffengy) [inline src/tls_stream.rs]: not wikipedia only, but yeah also.

File: src/tls_stream.rs (hunk 3 above)
Comment (steffengy) [inline src/tls_stream.rs]: Seems better, although it looks a bit strange. I'll add it

**Pull request merged**